infrastructure/constructs/bundled_code.py

```diff
@@ -1,14 +1,73 @@
-from os.path import join
+from pathlib import Path
+from re import sub
+from subprocess import PIPE, check_call, run
+from sys import executable
 
 from aws_cdk import BundlingOptions, aws_lambda
 
 from .backend import BACKEND_DIRECTORY
 from .lambda_config import PYTHON_RUNTIME
 
 
-def bundled_code(directory: str) -> aws_lambda.Code:
+def poetry_export_extras(packaging_directory: str, lambda_directory: str) -> None:
+    # There isn't an elegant way of getting poetry to install package dependencies in a bespoke
+    # target lambda_directory within Python, so we export a requirements file and install using pip.
+    # This has been raised and discussed by the community as below:
+    # https://github.com/python-poetry/poetry/issues/1937
+
+    export_extras = run(
+        ["poetry", "export", f"--extras={lambda_directory}", "--without-hashes"],
+        check=True,
+        stdout=PIPE,
+    ).stdout
+    # Remove botocore as this is already installed in the lambda layer
+    export_extras = sub(b"botocore==.*\n", b"", export_extras)
+
+    with open(
+        f"{packaging_directory}/{lambda_directory}/requirements.txt", "wb"
+    ) as lambda_requirements:
+        lambda_requirements.write(export_extras)
+
+
+def pip_install_requirements(packaging_directory: str, lambda_directory: str) -> None:
+    # Documentation recommend against calling pip internal api; rather, via command line
+    # https://pip.pypa.io/en/latest/user_guide/#using-pip-from-your-program
+
+    # Target already exists warning is discussed here
+    # https://github.com/pypa/pip/issues/8063
+    check_call(
+        [
+            executable,
+            "-m",
+            "pip",
+            "install",
+            "--no-deps",
+            "--quiet",
+            f"--cache-dir={packaging_directory}/cache",
+            f"--requirement={packaging_directory}/{lambda_directory}/requirements.txt",
```

Comment:
Can we pass the requirements as a string argument rather than a file? Then we won't need `requirements.txt` at all.

Also, `requirements.txt` should probably not be in the same directory we're using as an install target, since it'll be packaged with our Lambdas and include a file which is not needed.